File: code/ch11/11.3.2_characterReplacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        max_char_n = res = 0
        count = collections.Counter()
        for i in range(len(s)):
            count[s[i]] += 1
            max_char_n = max(max_char_n, count[s[i]])

            if res - max_char_n < k:
                res += 1
            else:
                count[s[i - res]] -= 1

        return res

File: code/ch11/11.3.2_characterReplacement.cpp
class Solution {
    public:
        int characterReplacement(string s, int k) {
            int res = 0;
            int maxf = 0;
            vector<int> count(26, 0);
            
            for (int i = 0; i < s.size(); i++) {
                count[s[i] - 'A'] += 1;
                maxf = max(maxf, count[s[i] - 'A']);
                
                if (res - maxf < k)
                    res += 1;
                else
                    count[s[i - res] - 'A'] -= 1;
            }

            return res;
        }
};
